README
WebService-Libris - a Perl module for retrieving book meta data from
libris.kb.se

INSTALLATION

To install this module, run the following commands:

	perl Makefile.PL
	make
	make test
	make install

SUPPORT AND DOCUMENTATION

After installing, you can find documentation for this module with the
perldoc command.

    perldoc WebService::Libris

You can also look for information at:

    Bug tracker:
        https://github.com/moritz/WebService-Libris/issues

    AnnoCPAN, Annotated CPAN documentation
        http://annocpan.org/dist/WebService-Libris

    CPAN Ratings
        http://cpanratings.perl.org/d/WebService-Libris

    Search CPAN
        http://search.cpan.org/dist/WebService-Libris/


LICENSE AND COPYRIGHT

Copyright (C) 2011 Moritz Lenz

This program is free software; you can redistribute it and/or modify it
under the terms of either: the GNU General Public License as published
by the Free Software Foundation; or the Artistic License.

See http://dev.perl.org/licenses/ for more information.



Changes
Revision history for WebService-Libris

0.07    2012-03-09
        Remove deprecated use of Mojo::DOM->new() with arguments
        Fixes failure reports by recent CPAN smokers

0.06
        Switch to Module::Build

0.05    2011-06-28
        Add missing FileCache.pm
        more elaborate language handling
        remove nonsensical -T from test shebang lines

0.03    2011-06-28
        Documentation for all interesting classes
        More accessors
        Caching
        Unit tests
        Got rid of WebService::Libris::Collection

0.02    2011-06-19
        Improved documentation
        Accessors are a bit more robust

0.01    2011-06-18
        First version, released on an unsuspecting world.


lib/WebService/Libris/Book.pm
package WebService::Libris::Book;
use Mojo::Base 'WebService::Libris';
use WebService::Libris::Utils qw/marc_lang_code_to_iso/;
use strict;
use warnings;
use 5.010;

__PACKAGE__->_make_text_accessor(qw/title date publisher/, ['isbn', 'isbn10']);

sub fragments {
    'bib', shift->id;
}

sub related_books { shift->list_from_dom('frbr_related') }
sub held_by       { shift->list_from_dom('held_by')      }
sub authors_obj   { shift->list_from_dom('creator')      }

sub authors_text  {
    my $self = shift;
    my @authors = grep length, map $_->text, $self->dom->find('creator')->each;
    # XXX: come up with something better
    if (wantarray) {
        return @authors;
    } elsif (@authors == 1) {
        return $authors[0]
    } else {
        return join ", ", @authors;
    }
}

sub isbns {
    my $self = shift;
    map $_->text, $self->dom->find('isbn10')->each;
}

sub authors_ids {
    my $self = shift;
    my %seen;
    my @ids = sort
              grep { !$seen{$_}++ }
              map { (split '/', $_)[-1] }
              grep $_,
              map { $_->attrs->{'rdf:resource'} }
              $self->dom->find('creator')->each;
    return @ids;
}

sub languages_marc {
    my $self = shift;

    my @l = $self->dom->find('language')->each;
    @l = grep $_, map $_->attrs->{'rdf:resource'}, @l;
    return undef unless @l;

    map { m{http://purl.org/NET/marccodes/languages/(\w{3})(?:\#lang)?} && "$1" } @l;
}

sub language_marc {
    (shift->languages_marc)[-1] // ()
}

sub languages {
    my $self = shift;
    my @langs = map marc_lang_code_to_iso($_), $self->languages_marc;
    for ($self->dom->find('*[lang]')->each) {
        my $l = $_->attrs->{'xml:lang'};
        push @langs, $l if defined $l;
    }
    @langs;
}

sub language {
    my $self = shift;
    my @langs = $self->languages;

    return undef unless @langs;
    my %c;
    ++$c{$_} for @langs;
    # just one language
    return $langs[0] if keys(%c) == 1;

    @langs = reverse sort { $c{$a} <=> $c{$b} } @langs;
    return $langs[0] if $c{$langs[0]} - $c{$langs[1]} >= 2;
    return undef;
}

=head1 NAME

WebService::Libris::Book - represents a Book in the libris.kb.se webservice

=head1 SYNOPSIS

    use WebService::Libris;
    for my $b (WebService::Libris->search(term => 'Rothfuss')) {
        # $b is a WebService::Libris::Book object here
        say $b->title;
        say $b->isbn;
    }

=head1 DESCRIPTION

C<WebService::Libris::Book> is a subclass of C<WebService::Libris> and
inherits all its methods.

All of the following methods return undef or the empty list if the information is not available.

=head1 METHODS

=head2 title

returns the title of the book

=head2 date

returns the publication date as a string (often just a year)

=head2 isbn

returns the first ISBN

=head2 isbn

returns a list of all ISBNs associated with the current book

=head2 publisher

returns the name of the publisher

=head2 related_books

returns a list of related books

=head2 held_by

returns a list of libraries that hold this book

=head2 authors_obj

returns a list of L<WebService::Libris::Author> objects which are listed
as I<creator> of this book.

=head2 authors_text

returns a list of creators of this book, as extracted from the response.
This often contains duplicates, or slightly different versions of the
same author name, so should be used with care.

=head2 language_marc

Returns the language in the three-letter "MARC" code, or undef if no such
code is found.

=head2 language

Some of the book records include a "MARC" language code (same as the
Library of Congress uses).  This methods tries to extract this code, and returns
the equivalent ISO 639 language code (two letters) if the translation is known.

It also exracts C<xml:lang> attribute from any tags found in the record.

Sometimes there are several different language specifications in a single
record.  In this case this method does an educated guess which one is correct.

=head2 languages

Return all language codes mentioned in the description of the C<language> method. No deduplication is done.


=cut

1;


lib/WebService/Libris/Author.pm
package WebService::Libris::Author;
use Mojo::Base 'WebService::Libris';
use strict;
use warnings;
use 5.010;

sub fragments {
    'auth', shift->id;
}

sub _description {
    my $self = shift;
    my $url = join '/', $self->fragments;
    $self->dom->at(qq{description[about\$="$url"]}) // Mojo::DOM->new;
}

sub birthyear {
    my $d = shift->_description->at('birthyear');
    $d && $d->text
}

sub libris_key {
    my $d = shift->_description->at('key');
    $d && $d->text
}

sub same_as {
    my $self = shift;
    my $sd = $self->_description->at('sameas');
    if ($sd) {
        return $sd->attrs->{'rdf:resource'};
    }
    return;
}

sub names {
    map $_->text, shift->_description->find('name')->each;
}

sub books {
    shift->list_from_dom('description[about^="http://libris.kb.se/resource/bib/"]');
}

=head1 NAME

WebService::Libris::Author - Author objects for WebService::Libris

=head1 SYNOSPIS

    use 5.010;
    use WebService::Libris;
    my $author = WebService::Libris->new(
        type    => 'autho',
        id      => '246603',
    );
    say $author->libris_key; # main name entry in the db
    for ($author->names) {
        say "    name variant: $_";
    }
    say "    identification URL: ", $author->same_as if $author->same_as;

=head1 DESCRIPTIONO

Author objects as returned from the libris.kb.se API search.

All of the following methods can return undef or the empty list if the
information is not available.

=head1 METHODS

C<WebService::Libris::Author> inherits from L<WebService::Libris>, and thus has all of its methods.

=head2 libris_key

Returns the canonical name of the author, as stored in libris.

=head2 names

Returns a list of alternative names used for that author. Often includes
spelling variations and translations into other languages.

=head2 birthyear

Returns the birth year of the author.

=head2 same_as

Returns an URL that uniquely identifies the author. Those URLs typically
point to viaf.org or dbpedia.org

=head2 books

Returns a list of books written by this author.

=cut

1;


lib/WebService/Libris/FileCache.pm
use 5.010;
package WebService::Libris::FileCache;
use Mojo::Base -base;
use Mojo::DOM;

has 'directory';

sub new {
    my $class = shift;
    my $self = bless {@_}, $class;
    $self->_init;
    $self
}

sub _init {
    my $self = shift;
    my $d = $self->directory;
    unless (-d $d) {
        require File::Path;
        File::Path::make_path($d)
    }
}

sub _filename {
    my ($self, $key) = @_;
    $key =~ s{/}{_}g;
    $self->directory . $key . '.xml'
}

sub get {
    my ($self, $key) = @_;
    my $filename = $self->_filename($key);
    return undef unless open my $h, '<', $filename;
    my $contents = do { local $/; <$h> };
    return undef unless length $contents;
    Mojo::DOM->new->xml(1)->charset('UTF-8')->parse($contents);
}

sub set {
    my ($self, $key, $value) = @_;
    my $filename = $self->_filename($key);
    open my $h, '>', $filename or die "Can't open file '$filename' for writing: $!";
    print { $h } $value->to_xml;
    close $h;
    $value;
}


1;


lib/WebService/Libris/Collection.pm
package WebService::Libris::Collection;
use Mojo::Base -base;

require WebService::Libris;

has 'type';
has 'ids';
has 'cache';

sub new {
    my ($class, %attrs) = @_;
    bless \%attrs, $class;
}

sub all {
    my $self = shift;
    map WebService::Libris->new(type => $self->type, id => $_, cache => $self->cache),
        @{ $self->ids };
}

sub first {
    my $self = shift;
    return unless @{ $self->ids };
    WebService::Libris->new(
        type    => $self->type,
        id      => $self->ids->[0],
        cache   => $self->cache,
    );
}

sub next {
    my $self = shift;
    if (@{ $self->ids }) {
        return WebService::Libris->new(
            type    => $self->type,
            id      => shift @{ $self->ids },
            cache   => $self->cache,
        )
    } else {
        return;
    }
}

sub elems {
    scalar @{ shift->ids }
}

=head1 NAME

WebService::Libris::Collection - iterable collection of objects

=head1 SYNOPSIS

    use WebService::Libris;
    my $collection = WebService::Libris->search(term => 'blomkvist');
    while (my $book = $collection->next) {
        # do something with $book
    }

=head1 DESCRIPTION

WebService::Libris::Collection objects hold list of objects of subclasses of
L<WebService::Libris>. It is used instead of arrays because it delays requests
to the libris.kb.se web API until an object is actually access.

=head1 METHODS

=head2 first

Returns the first item

=head2 next

Returns the next item, and removes it from the internal list of items, so that
repeated calls to C<next> iterate over the whole collection

=head2 all

Returns a list of all remaining objects in the collection

=head2 elems

Returns the number of remaining objects in the collection

=cut

1;


lib/WebService/Libris.pm
package WebService::Libris;
use Mojo::Base -base;
use Mojo::UserAgent;
use Mojo::URL;

use 5.010;
use strict;
use warnings;

my %default_typemap = (
    bib     => 'Book',
    book    => 'Book',
    auth    => 'Author',
    author  => 'Author',
    library => 'Library',
);

has 'id';
has 'type';
has '_dom';
has 'cache';

has 'type_map';

=head1 NAME

WebService::Libris - Access book meta data from libris.kb.se

=head1 VERSION

Version 0.07

Note that the API is still subject to change.

=cut

our $VERSION = '0.07';


=head1 SYNOPSIS

    use WebService::Libris;
    use 5.010;
    binmode STDOUT, ':encoding(UTF-8)';

    my $book = WebService::Libris->new(
        type => 'book',
        # Libris ID
        id   => '9604288',
        # optional but recommended:
        cache_dir = '/tmp/webservice-libris/',
    );
    print $book->title;

    my $books = WebService::Libris->search(
        term    => 'Astrid Lindgren',
        page    => 1,
    );
    while (my $b = $books->next) {
        say $b->title;
        say '  isbn: ', $b->isbn;
        say '  date: ', $b->date;
    }

=head1 DESCRIPTION

The Swedish public libraries and the national library of Sweden have a common
catalogue containing meta data of the books they have available.

This includes many contemporary as well as historical books.

The catalogue is available online at L<http://libris.kb.se>, and can be
queried with a public API.

This module is a wrapper around two of their APIs (xsearch and RDF responses).

=head1 METHODS

=head2 new

    my $obj = WebService::Libris->new(
        type => 'author',
        id   => '246603',
    );

Creates an object of the C<WebService::Libris> class or a subclass thereof
(denoted by C<type> in the argument list). C<type> can currently be one of
(synonyms on one line)

    auth author
    bib book
    library

The C<id> argument is mandatory, and must contain the Libris ID of the object
you want to retrieve. If you don't know the Libris ID, use one of the
C<search> functions instead.

=head2 direct_search

    my $hashref = WebService::Libris->direct_search(
        term    => 'Your Searchterms Here',
        page    => 1,   # page size is 200
        full    => 1,   # return all available information
    );

Returns a hashref directly from the JSON response of the xsearch API
described at L<http://librishelp.libris.kb.se/help/xsearch_eng.jsp?open=tech>.

This is more efficient than a C<< WebService::Libris->search >> call, because
it does only one query (whereas C<< ->search >> does one additional request
per result object), but it's not as convenient, and does not allow browsing of
related entities (such as authors and libraries).

=head2 search

    my @books = WebService::Libris->search(
        term    => 'Your Search Term Here',
        page    => 1,
    );
    for my $book (@books) {
        say $book->title;
    }

Searches the xsearch API for arbitrary search terms, and returns a
C<WebService::Libris::Collection> of books.

See the C<direct_search> method above for a short discussion.

=head2 search_for_isbn

    my $book = WebService::Libris->search_for_isbn('9170370192');

Looks up a book by ISBN

=head1 Less interesting methods

The following methods aren't usually useful for the casual user, more
for those who want to extend or subclass this module.

=head2 rdf_url

Returns the RDF resource URL for the current object. Mostly useful for internal purposes.

=head2 dom

Returns the L<Mojo::DOM> object from the web services response.
Does a request to the web service if no DOM was stored previously.

Only useful for you if you want to extract more data from a response
than the object itself provides.

=head2 id

Returns the libris ID of the object. Only makes sense for subclasses.

=head2 type

Returns the short type name (C<bib>, C<auth>, C<library>). Only makes sense
for subclasses.

=head2 fragments

Must be overridden in a subclass to return a list of
the last two junks of the RDF resource URL, that is the short
type name and the libris ID.

=head1 AUTHOR

Moritz Lenz, C<< <moritz at faui2k3.org> >>

=head1 BUGS

Please report any bugs or feature requests at
L<https://github.com/moritz/WebService-Libris/issues>

=head1 SUPPORT

You can find documentation for this module with the perldoc command.

    perldoc WebService::Libris

You can also look for information at:

=over 4

=item * Bug tracker:

L<https://github.com/moritz/WebService-Libris/issues>

=item * AnnoCPAN: Annotated CPAN documentation

L<http://annocpan.org/dist/WebService-Libris>

=item * CPAN Ratings

L<http://cpanratings.perl.org/d/WebService-Libris>

=item * Search CPAN

L<http://search.cpan.org/dist/WebService-Libris/>

=back

=head1 BUGS AND LIMITATIONS

Nearly no error checking is done. So beware!

=head1 ACKNOWLEDGEMENTS

Thanks go to the Kungliga biblioteket (National Library of Sweden) for
providing the libris.kb.se service and API.

=head1 LICENSE AND COPYRIGHT

Copyright 2011 Moritz Lenz.

This program is free software; you can redistribute it and/or modify it
under the terms of either: the GNU General Public License as published
by the Free Software Foundation; or the Artistic License.

See L<http://dev.perl.org/licenses/> for more information.

=cut

sub new {
    my ($class, %opts) = @_;;
    my $c;
    if ($opts{type}) {
        if ($opts{type_map}) {
            $c = $opts{type_map}{lc $opts{type}}
                // $default_typemap{lc $opts{type}};
        } else {
            $c = $default_typemap{lc $opts{type}};
        }
    }
    if (my $cache_dir = delete $opts{cache_dir}) {
        require WebService::Libris::FileCache;
        $opts{cache} = WebService::Libris::FileCache->new(
            directory => $cache_dir,
        );
    }
    if ($c) {
        $class = __PACKAGE__ . "::" . $c;
        eval "use $class; 1" or die $@;
        return bless \%opts, $class;
    } else {
        return bless \%opts, $class;
    }
}


sub rdf_url {
    my $self = shift;
    my ($key, $id) = $self->fragments;
    "http://libris.kb.se/data/$key/$id?format=application%2Frdf%2Bxml";
}

sub dom {
    my $self = shift;

    unless ($self->_dom) {
        if ($self->cache) {
            my $key  = join '/', $self->fragments;
            if (my $r = $self->cache->get($key)) {
                $self->_dom($r);
            } else {
               my $dom = $self->_request_dom;
               $self->cache->set($key, $dom);
               $self->_dom($dom);
            }
        } else {
            $self->_dom($self->_request_dom);
        }
    }
    $self->_dom;
}

sub _request_dom {
    my $self = shift;
    Mojo::UserAgent->new()->get($self->rdf_url)->res->dom;
}

sub direct_search {
    my ($self, %opts) = @_;
    my $terms = $opts{term} // die "Search term missing";
    my $page  = $opts{page} // 1;
    my %q = (
        query   => $terms,
        n       => 200,     # max. number of results
        start   => 1 + 200 * ($page - 1),
        format  => 'json',
    );
    $q{format_level} = 'full' if $opts{full};
    my $url = Mojo::URL->new('http://libris.kb.se/xsearch');
    $url->query(%q);
    my $res = Mojo::UserAgent->new()->get($url)->res;
    $res->json;
}

sub search {
    my ($self, %opts) = @_;
    my $json = $self->direct_search(%opts);
    my @ids = map { (split '/',  $_->{identifier})[-1] }
                  @{ $json->{xsearch}{list} };
    WebService::Libris::Collection->new(
        type    => 'bib',
        ids     => \@ids,
        cache   => $self->cache,
    );
}

sub search_for_isbn {
    my ($self, $isbn) = @_;
    my $res = Mojo::UserAgent->new->max_redirects(1)
              ->get("http://libris.kb.se/hitlist?q=linkisxn:$isbn");
    my $url = $res->res->headers->location;
    return unless $url;
    my ($type, $libris_id) = (split '/', $url)[-2, -1];
    $self->new(type => $type, id => $libris_id, cache => $self->cache);
}

sub fragments {
    die "Must be overridden in subclasses";
}

sub list_from_dom {
    my ($self, $search_for) = @_;
    my $key;
    my @result;
    my %seen;
    $self->dom->find($search_for)->each(sub {
        my $d = shift;
        my $resource_url =  $d->attrs->{'rdf:resource'}
                         // $d->attrs->{'rdf:about'};
        return unless $resource_url;
        my ($k, $id) = $self->fragment_from_resource_url($resource_url);
        return if $seen{"$k/$id"}++;
        push @result, __PACKAGE__->new(
            type    => $k,
            id      => $id,
            cache   => $self->cache,
        );
    });
    @result;
}

sub fragment_from_resource_url {
    my ($self, $url) = @_;
    (split '/', $url)[-2, -1];
}

sub _make_text_accessor {
    my $package = shift;
    for (@_) {
        my ($name, $look_for);
        if (ref($_) eq 'ARRAY') {
            ($name, $look_for) = @$_;
        } else {
            $name     = $_;
            $look_for = $_;
        }
        no strict 'refs';
        *{"${package}::$name"} = sub {
            my $thing;
            ($thing = shift->dom->at($look_for)) && $thing->text;
        };
    }
}


1; # End of WebService::Libris
